src/main/java/org/prebid/server/bidder/ogury/OguryBidder.java

```diff
@@ -0,0 +1,277 @@
+        if (markupType == null) {
+            throw new PreBidException("Missing MType for impression: `%s`".formatted(bid.getImpid()));
+        }
+
+        return switch (markupType) {
+            case 1 -> BidType.banner;
+            case 2 -> BidType.video;
+            case 3 -> BidType.audio;
+            case 4 -> BidType.xNative;
+            default -> throw new PreBidException(
+                    "Unsupported MType '%d', for impression '%s'".formatted(bid.getMtype(), bid.getImpid()));
+        };
+    }
+
+    private BidderError validateRequestKeys(BidRequest request, List<Imp> impsWithOguryParams) {
+        final Optional<Site> siteOpt = Optional.of(request).map(BidRequest::getSite);
+        final Optional<String> publisherId = siteOpt.map(Site::getPublisher).map(Publisher::getId);
+
+        if (CollectionUtils.isEmpty(impsWithOguryParams) && (siteOpt.isEmpty() || publisherId.isEmpty())) {
+            return BidderError.badInput("Invalid request. assetKey/adUnitId or request.site.publisher.id required");
+        }
+
+        return null;
+    }
+
+    private BidderError checkResponseStatusCodeForErrors(HttpResponse response) {
+        final int statusCode = response.getStatusCode();
+
+        if (statusCode == HttpResponseStatus.BAD_REQUEST.code()) {
+            return BidderError.badInput("Unexpected status code: %d. Run with request.debug = 1 for more info"
+                    .formatted(statusCode));
+        }
+
+        if (statusCode != HttpResponseStatus.OK.code()) {
+            return BidderError.generic("Unexpected status code: %d. Run with request.debug = 1 for more info"
+                    .formatted(statusCode));
+        }
+
+        return null;
+    }
+
+    private boolean isNotHasContent(HttpResponse response) {
+        return Optional.of(response)
+                .map(HttpResponse::getStatusCode)
+                .map(code -> code == HttpResponseStatus.NO_CONTENT.code() || StringUtils.isEmpty(response.getBody()))
+                .orElse(false);
+    }
+
+    private MultiMap buildHeaders(BidRequest request) {
+        final MultiMap headers = HttpUtil.headers();
+
+        Optional.ofNullable(request)
+                .map(BidRequest::getDevice)
+                .ifPresentOrElse(device -> {
+                    final String lang = device.getLanguage();
+                    headers.add(HttpUtil.USER_AGENT_HEADER, device.getUa())
+                            .add(HttpUtil.ACCEPT_LANGUAGE_HEADER, lang != null ? lang : "en-US");
+
+                    Optional.of(device)
+                            .map(Device::getIp)
+                            .ifPresent(ip -> headers.add(HttpUtil.X_FORWARDED_FOR_HEADER, ip));
+
+                    Optional.of(device)
+                            .map(Device::getIpv6)
+                            .ifPresent(ip -> headers.add(HttpUtil.X_FORWARDED_FOR_HEADER, ip));
+                }, () -> headers.add(HttpUtil.ACCEPT_LANGUAGE_HEADER, "en-US"));
```

Comment:
Please, see other examples how to do it properly. For example `KrushmediaBidder`